src/session/sidebar/avatar.rs

```diff
@@ -0,0 +1,265 @@
+use gtk::prelude::*;
+use gtk::subclass::prelude::*;
+use gtk::{glib, gsk};
+use tdgrand::enums::{ChatType, UserStatus, UserType};
+
+use crate::session::user::BoxedUserStatus;
+use crate::session::{Chat, Session, User};
+
+mod imp {
+
+    use super::*;
+    use gtk::CompositeTemplate;
+    use std::cell::{Cell, RefCell};
+
+    use crate::session::components::Avatar as ComponentsAvatar;
+
+    #[derive(Debug, Default, CompositeTemplate)]
+    #[template(resource = "/com/github/melix99/telegrand/ui/sidebar-avatar.ui")]
+    pub struct Avatar {
+        pub chat: RefCell<Option<Chat>>,
+        pub is_online: Cell<bool>,
+        // The first Option indicates whether we've once tried to compile the shader. The second
```

Comment:
The first option just indicates whether we once have tried to compile the shader and to prevent it for being recompiled (no matter if the compilation failed or was successful). 
The original C code had a boolean flag for that, but this is a more idiomatic way in rust, I think.
The second Option indicates whether the shader compilation was actually successful (Then there is a some shader). 
Maybe it will be clearer when you look at line 236 and 255.

Reply:
> (no matter if the compilation failed or was successful).

That was the missing bit for me, I understand now.